packages/ts/hilla-file-router/src/generateRoutes.ts

```diff
@@ -0,0 +1,94 @@
+import { extname, relative } from 'node:path';
+import { fileURLToPath } from 'node:url';
+import { template, transform as transformer } from '@vaadin/hilla-generator-utils/ast.js';
+import createSourceFile from '@vaadin/hilla-generator-utils/createSourceFile.js';
+import ts, {
+  type ImportDeclaration,
+  type ObjectLiteralExpression,
+  type StringLiteral,
+  type VariableStatement,
+} from 'typescript';
+import type { RouteMeta } from './collectRoutes.js';
+import { processPattern, transformRoute } from './utils.js';
+
+const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
+
+function relativize(url: URL, generatedDir: URL): string {
+  const result = relative(fileURLToPath(generatedDir), fileURLToPath(url));
+
+  if (!result.startsWith('.')) {
+    return `./${result}`;
+  }
+
+  return result;
+}
+
+function createImport(mod: string, file: string): ImportDeclaration {
+  const path = `${file.substring(0, file.lastIndexOf('.'))}.js`;
+  return template(`import * as ${mod} from '${path}';\n`, ([statement]) => statement as ts.ImportDeclaration);
+}
+
+function createRouteData(
+  path: string,
+  mod: string | undefined,
+  children: readonly ObjectLiteralExpression[],
+): ObjectLiteralExpression {
+  return template(
+    `const route = {
+  path: '${path}',
+  ${mod ? `module: ${mod}` : ''}
+  ${children.length > 0 ? `children: CHILDREN,` : ''}
+}`,
+    ([statement]) =>
+      (statement as VariableStatement).declarationList.declarations[0].initializer as ObjectLiteralExpression,
+    [
+      transformer((node) =>
+        ts.isIdentifier(node) && node.text === 'CHILDREN' ? ts.factory.createArrayLiteralExpression(children) : node,
+      ),
+    ],
+  );
+}
+
+export default function generateRoutes(views: RouteMeta, generatedDir: URL): string {
+  const imports: ImportDeclaration[] = [];
+  let id = 0;
+
+  const routes = transformRoute<RouteMeta, ObjectLiteralExpression>(
+    views,
+    (view) => view.children.values(),
+    ({ file, layout, path }, children) => {
+      const currentId = id;
+      id += 1;
+
+      let mod: string | undefined;
+      if (file) {
+        mod = `Page${currentId}`;
+        imports.push(createImport(mod, relativize(file, generatedDir)));
+      } else if (layout) {
+        mod = `Layout${currentId}`;
+        imports.push(createImport(mod, relativize(layout, generatedDir)));
+      }
+
+      return createRouteData(processPattern(path), mod, children);
+    },
+  );
+
+  const routeDeclaration = template(
+    `import a from 'IMPORTS';
+
+const routes = ROUTE;
+
+export default routes;
+`,
+    [
+      transformer((node) =>
+        ts.isImportDeclaration(node) && (node.moduleSpecifier as StringLiteral).text === 'IMPORTS' ? imports : node,
+      ),
+      transformer((node) => (ts.isIdentifier(node) && node.text === 'ROUTE' ? routes : node)),
+    ],
+  );
+
+  const file = createSourceFile(routeDeclaration, 'views.ts');
+
+  return printer.printFile(file);
```

Comment:
Just a question:
Why the view routes are being persisted to a file here, and then in the vite plugin they are being read and written to another file? Couldn't the transformed `routeDeclaration` be returned and persisted there once?

Reply:
Yeah, that looks a bit misleading here but that's more like a problem of Typescript API naming. The `printFile` name here means that the Typescript prints a structure called `SourceFile` which basically contains the content of the whole file AST. It has pretty thin relation to the real file.